File: tests/xmp-api/global-view/F/Laplace/xmp_laplace_1d.f90
program laplace
  integer,parameter :: N1=100, N2=200
  real(8),parameter :: PI = 3.141592653589793238463

  integer :: niter=10
  real(8) :: u(N1,N2),uu(N1,N2)
  real(8) :: value = 0.0

!$xmp nodes p(*)
!$xmp template t(N1,N2)
!$xmp distribute t(*, block) onto p
!$xmp align u(i,j) with t(i, j)
!$xmp align uu(i,j) with t(i, j)
!$xmp shadow uu(0:0,1:1)

!$xmp loop (i,j) on t(i,j)
  do j=1,N2
     do i=1,N1
        u(i,j)=0.0
        uu(i,j)=0.0
     end do
  end do

!$xmp loop (i,j) on t(i,j)
  do j=2,N2-1
     do i=2,N1-1
        u(i,j)=sin(dble(i-1)/N1*PI)+cos(dble(j-1)/N2*PI)
     end do
  end do

  do k=1,niter

     if(xmp_node_num() == 1) then
        print *,'k=',k
     end if
!$xmp loop (i,j) on t(i,j)
     do j=2,N2-1
        do i=2,N1-1
           uu(i,j)=u(i,j)
        end do
     end do

!$xmp reflect (uu)

!$xmp loop (i,j) on t(i,j)
     do j=2,N2-1
        do i=2,N1-1
           u(i,j)=(uu(i-1,j) + uu(i+1,j) + uu(i,j-1) + uu(i,j+1))/4.0
        end do
     end do


     value = 0.0
!$xmp loop (i,j) on t(i,j) reduction(+:value)
     do j=2,N2-1
        do i=2,N1-1
           value = value + dabs(uu(i,j) -u(i,j))
        end do
     end do

!! !$xmp task on p(1,1)
     if(xmp_node_num() == 1) then
        print *, 'Verification =', value
     end if
!! !$xmp end task

  enddo
  
end program laplace

File: tests/xmp-api/global-view/C/Laplace/xmp_laplace_1d.c
#include <stdio.h> 
#include <stdlib.h> 
#include <math.h>
#include <xmp.h>

#define N1 100
#define N2 200
double u[N2][N1], uu[N2][N1];

#pragma xmp nodes p[*]
#pragma xmp template t[N2][N1]
#pragma xmp distribute t[block][*] onto p
#pragma xmp align u[j][i] with t[j][i]
#pragma xmp align uu[j][i] with t[j][i]
#pragma xmp shadow uu[1:1][0:0]

int main(int argc, char **argv)
{
  int i, j, k, niter = 10;
  double value = 0.0;

#pragma xmp loop (j,i) on t[j][i]
  for(j = 0; j < N2; j++){
    for(i = 0; i < N1; i++){
      u[j][i] = 0.0;
      uu[j][i] = 0.0;
    }
  }

#pragma xmp bcast(value)

#pragma xmp loop (j,i) on t[j][i]
  for(j = 1; j < N2-1; j++)
    for(i = 1; i < N1-1; i++)
      u[j][i] = sin((double)i/N1*M_PI) + cos((double)j/N2*M_PI);
  
  for(k = 0; k < niter; k++){

    // if(xmpc_all_node_num() == 0) printf("iter =%d\n",k);

#pragma xmp loop (j,i) on t[j][i]
    for(j = 1; j < N2-1; j++)
      for(i = 1; i < N1-1; i++)
	uu[j][i] = u[j][i];
    
#pragma xmp reflect (uu)

#pragma xmp loop (j,i) on t[j][i]
    for(j = 1; j < N2-1; j++)
      for(i = 1; i < N1-1; i++)
	u[j][i] = (uu[j-1][i] + uu[j+1][i] + uu[j][i-1] + uu[j][i+1])/4.0;
  
    value = 0.0;
#pragma xmp loop (j,i) on t[j][i] reduction(+:value)
    for(j = 1; j < N2-1; j++)
      for(i = 1; i < N1-1; i++)
	value += fabs(uu[j][i] - u[j][i]);
    //#pragma xmp task on p[0][0]
    if(xmpc_all_node_num() == 0)
      printf("k=%d: Verification = %20.16f\n", k, value);
  }

  return 0;
}
